Add history_values sensor support

mvn23 opened the pull request:
## Description:
Inspired by the history_stats sensor, this sensor can take numeric values from the history and calculate things like min/max, average and range. Useful for e.g. energy consumption or solar panel yield summaries.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7580

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
- platform: history_values
  name: "Solar Panel Yield Yesterday"
  type: range
  target_entity_id: sensor.total_yield
  start: "{% set today %}{{ now().year }}-{{ now().month }}-{{ now().day }}{% endset %}{{ as_timestamp(strptime(today, '%Y-%m-%d')) - 86400 }}"
  end: "{% set today %}{{ now().year }}-{{ now().month }}-{{ now().day }}{% endset %}{{ as_timestamp(strptime(today, '%Y-%m-%d')) }}"
  update:
    at: '00:00:01'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - N/A

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

mvn23 committed:
Add history_values sensor support

mvn23 committed:
Fix precision parsing

dgomes commented:
Too many options to set start end... just use cv.time

dgomes commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+            str(self._target_entity))
+        history_list = history_states.get(self._target_entity)
+        if history_list is None:
+            self.value = STATE_UNKNOWN
```
don't use STATE_UNKNOWN use None

dgomes replied:
should be in developers.home-assistant.io

mvn23 replied:
I haven't been able to find it there...

dgomes commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+        total_period = datetime.timedelta(seconds=0)
+        last_idx = len(state_list)-1
+        for state in state_list:
+            if state.state == STATE_UNKNOWN:
```
start by filtering out the STATE_UNKNOWNS:

state_list = [state for state in state_list if state.state != STATE_UNKNOWN]

mvn23 replied:
Addressed in c89c09c by removing all non-numeric states from the list before processing.

dgomes commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+            if state.state == STATE_UNKNOWN:
+                # Don't take this into account at all.
+                continue
+            current_prec = self.get_precision(float(state.state))
```
find the largest precision:

```precision = max(state_list, lambda state: self.get_precision(float(state.state)))

mvn23 replied:
addressed in c89c09c

dgomes reviewed (commented):
I think this sensor meant to be generic (beyond energy meter), but it is not...

I would advise making it a true energy meter (and remove over generalisations)

mvn23 commented:
> Too many options to set start end... just use cv.time

The templates for start and end allow for flexibility. cv.time would only allow to specify a time of day, not a point in time (e.g. start of yesterday). Apart from this, this part of the config was more or less copied from sensor.history_stats, keeping it identical for consistency makes sense to me.

> I think this sensor meant to be generic (beyond energy meter), but it is not...
I would advise making it a true energy meter (and remove over generalisations)

The idea started for energy meters indeed, but it can be used for all types of numeric values. Think  temperature, humidity, pressure, but also everything that can be connected to a counter, like custom automations.

As for your other comments, there are definitely some valid points that I was not aware of or had not thought of. Expect new commits soon.

dgomes commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+        return self._unit_of_measurement
+
+    @callback
+    def get_average(self, state_list, end, now):
```
your absolutely correct: it's a **time-weighted** average (different things and we should be precise in respect to math concepts)

mvn23 replied:
Updated the docstring in 1503d4

mvn23 committed:
Rewrite get_min_max_range to Decimal()
Clarify docstring for get_average()
Filter history_list before passing it to subsequent functions

mvn23 commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+                next_time = state_list[next_updated].last_changed
+            state_period = next_time - state.last_changed
+            total_period += state_period
+            total_value += float(state.state) * state_period.total_seconds()
```
Looking into this now, but there does not seem to be a way to reduce() both total_period and total_value without calculating state_period multiple times for each state (once per reduce() call) or storing it somewhere for later use. Am I missing something?

dgomes replied:
I've been thinking about this sensor... why don't you go directly to the RDBMS to get these values ? You require the aggregate values, but transfer lots of information out of the RDBMS...

Could an SQL query get the information you need ?

mvn23 replied:
Let me start out by saying that I don't know exactly how the data is stored in the database. From my experience with SQL so far I can imagine that the `state_period`s would still need to be calculated and added up outside of the DB engine. `total_value` is a sum of the product of each state and its duration and since the duration would probably need to be calculated outside of the DB engine, so does this.
Apart from the issues above, I personally would not want to maintain a project which has DB access and SQL code scattered all over the place. I would prefer to keep all of that stuff in a separate module to make changing/upgrading the database schema in the future easier.

mvn23 replied:
Looked into reduce(), found no way to make significant improvements.

mvn23 committed:
Remove all non-numeric states from the list before processing.
Order imports
Change the way precision is determined

mvn23 commented on homeassistant/components/sensor/history_values.py:
```diff
@@ -0,0 +1,369 @@
+        if total_period.total_seconds() == 0:
+            return None
+        formatstr = '{:.' + str(precision) + 'f}'
+        return formatstr.format(total_value / total_period.total_seconds())
```
Changed the determination of min/max/range values to Decimal in 1503d4.
Decimal does not offer advantages for calculation of the time-weighted average.

mvn23 committed:
Change error to warning for non-numeric values in history

mvn23 committed:
Fix CI complaints

balloob commented:
I don't think that we should try to recreate SQL in the form of multiple similar but slightly different sensors to extract historical data.

balloob commented:
I also don't understand why this wouldn't be added as a different mode to the history_stats sensor. This should not be yet another sensor.

mvn23 commented:
> I also don't understand why this wouldn't be added as a different mode to the history_stats sensor. This should not be yet another sensor.

The reason I made it a separate sensor is that history_stats deals with binary states or the binary match of a state against a certain value. This sensor only deals with numerical states. I'm sure they could be merged, but then the config of this combined sensor would be rather extended and have a lot of conditional requirements (e.g. if the target entity is non-numeric, certain types of the sensor don't make sense. If it is numeric, these types only work if you don't specify a match, etc.)
In my opinion the only thing this sensor has in common with the history_stats sensor is the fact that both sensors fetch data from the history over a certain timespan.

balloob commented:
If it's just for numeric states, we should make sure the name represents that.

mvn23 commented:
> If it's just for numeric states, we should make sure the name represents that.

`history_numbers` maybe?

balloob commented:
I'm thinking that it might be better to just merge the 2 sensors  and have them have different files and pick implementation based on configuration.

```
components/sensor/history_stats/__init__.py
components/sensor/history_stats/numeric.py
components/sensor/history_stats/binary.py
```

We do the same thing for light.mqtt.
